docs/src/starknet/calldata-transformation.md

```diff
@@ -0,0 +1,206 @@
+# Calldata Transformation
+
+For the examples below, we will consider a dedicated contract - `DataTransformerContract`, defined in `data_transformer_contract` project namespace.
+
+It's declared on Sepolia network with class hash `0x02a9b456118a86070a8c116c41b02e490f3dcc9db3cad945b4e9a7fd7cec9168`.
+
+It has a few methods accepting different types and items defined in its namespace:
+
+```rust
+// data_transformer_contract/src/lib.cairo
+
+pub struct SimpleStruct {
+    a: felt252
+}
+
+pub struct NestedStructWithField {
+    a: SimpleStruct,
+    b: felt252
+}
+
+pub enum Enum {
+    One: (),
+    Two: u128,
+    Three: NestedStructWithField
+}
+
+#[starknet::contract]
+pub mod DataTransformerContract {
+    /* ... */
+
+    use super::*;
+
+    fn tuple_fn(self: @ContractState, a: (felt252, u8, Enum)) { ... }
+
+    fn complex_struct_fn(self: @ContractState, a: ComplexStruct) { ... }
+
+    fn complex_fn(
+        self: @ContractState,
+        arr: Array<Array<felt252>>,
+        one: u8,
+        two: i16,
+        three: ByteArray,
+        four: (felt252, u32),
+        five: bool,
+        six: u256
+    ) {
+        ...
+    }
+}
+```
+
+A default form of calldata passed to commands requiring it is a series of hex-encoded felts:
+
+```shell
+$ sncast --account myuser \
+    call \
+    --url http://127.0.0.1:5050 \
+    --contract-address 0x016ad425af4585102e139d4fb2c76ce786d1aaa1cfcd88a51f3ed66601b23cdd \
+    --function tuple_fn \
+    --calldata 0x10 0x3 0x0 \
+    --block-id latest
+```
+
+However, `sncast` allows passing the data in far more handy, human-readable form - as a series of Cairo expressions.
+When calldata is delivered in such form, Cast will perform serialization automatically, based on an ABI of the contract we interact with.
+
+To transform Cairo expressions to serialized form, calldata must meet the following requirements:
+
+* Arguments are comma-separated (on a contrary to a standard serialized input which is whitespace-separatd). Note the lack of commas in the example above - it signals Cast that the input is serialized.
+
+* If only one argument without a comma is given, it should be a *nontrivial* literal (e.g. a struct or enum constructor).
+  It cannot be a simple numeric litral - `--calldata 0x2137` will be treated as serialized.\
+  To transform such, we need to add a type suffix - for example: `--calldata 0x2137_u96`. See [Caveats](./calldata-transformation.md#caveats)
```

Comment:
Would a trailing comma after expression work as well?

Reply:
Yeah, trailing comma works. I modified one of the test cases to verify it.